src/ZPublisher/cookie.py

```diff
@@ -0,0 +1,337 @@
+
+# ``Expires``
+class UTC(datetime.tzinfo):
+    ZERO = datetime.timedelta(0)
+
+    def utcoffset(self, dt):
+        return self.ZERO
+
+    def tzname(self, dt):
+        return "UTC"
+
+    dst = utcoffset
+
+
+utc = UTC()
+
+wdmap = dict(enumerate("Mon Tue Wed Thu Fri Sat Sun".split()))
+mmap = dict((i + 1, m) for (i, m) in enumerate(
+    "Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec".split()))
+
+
+def rfc1123_converter(value):
+    """convert *value* into an RFC1123 date.
+
+    *value* can be a string (assumed already in the required form),
+    a Python `datetime` (assumed to be in the UTC time zone, if naive),
+    a float (interpreted as time stamp),
+    or a `DateTime` object.
+    """
+    if isinstance(value, str):
+        return value
+    if isinstance(value, (int, float)):
+        value = datetime.datetime.utcfromtimestamp(value)
+    elif isinstance(value, DateTime):
+        value = value.toZone("GMT").asdatetime()
+    if not isinstance(value, datetime.datetime):
+        raise ValueError("unsupported type: %s" % type(value))
+    if value.utcoffset() is not None:
+        value = value.astimezone(utc)
+    tt = value.timetuple()
+    # we cannot use ``strftime`` because it depends on locale
+    return "%s, %02d %s %04d %02d:%02d:%02d GMT" % (
+        wdmap[tt.tm_wday],
+        tt.tm_mday, mmap[tt.tm_mon], tt.tm_year,
+        tt.tm_hour, tt.tm_min, tt.tm_sec)
+
+
+registerCookieParameter("Expires", rfc1123_converter)
+
+
+# ``Max-Age``
+def int_converter(value):
+    """check for *int* value."""
+    int(value)
+    return str(value)
+
+
+registerCookieParameter("Max-Age", int_converter)
+
+
+# ``Domain``
+def domain_converter(value):
+    """convert *value* into an internationalized domain name.
+
+    Note: The Python infrastructure supports IDNA2003, but
+    RFC6265 calls for IDNA2008. IDNA2008 is implemented by
+    a third party module. You might want to override
+    the registration to get IDNA2008 if you observe domain
+    related problems.
+    """
+    if isinstance(value, bytes):
+        value = value.decode("utf-8")
+    if "xn--" in value:  # already encoded
+        # ATT: under Python 2, this returns unicode
+        # this may pose problems should other parameters
+        # provide encoded non-ASCII
+        return value
+    from encodings.idna import ToASCII
+    from encodings.idna import nameprep
+    return ".".join(to_str(ToASCII(nameprep(c))) for c in value.split("."))
+
+
+registerCookieParameter("Domain", domain_converter)
+
+
+# ``Path``
+# we include "%" to handle the case that the path has already been quoted
+path_safe = compile(r"[a-zA-Z0-9_/%]*$").match
+
+
+def path_converter(value):
+    """convert *value* to a path.
+
+    The convertion is based on ``absolute_url_path``.
+    If *value* lacks this method, it is assumed to be a string.
+    If the string contains ``%``, it is assumed that it is already
+    quoted.
+    """
+    ap = getattr(value, "absolute_url_path", None)
```

Comment:
If value is a Zope object it may acquire `absolute_url_path` if it doesn't have it, leading to a bad path. This is unlikely but possible.

Reply:
I will check for this.